### PR TITLE
KAFKA-17393: Remove log.message.format.version/message.format.version (KIP-724)

### PR DESCRIPTION
**Pull request by FrankYang0529**
Based on [KIP-724](https://cwiki.apache.org/confluence/display/KAFKA/KIP-724%3A+Drop+support+for+message+formats+v0+and+v1), the `log.message.format.version` and `message.format.version` can be removed in 4.0.

These configs effectively a no-op with inter-broker protocol version 3.0 or higher
since Apache Kafka 3.0, so the impact should be minimal.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/ConfigHandler.scala [hunk 1]
```diff
@@ -60,11 +57,9 @@ class TopicConfigHandler(private val replicaManager: ReplicaManager,
   private def updateLogConfig(topic: String,
                               topicConfig: Properties): Unit = {
     val logManager = replicaManager.logManager
-    // Validate the configurations.
-    val configNamesToExclude = excludedConfigs(topic, topicConfig)
     val props = new Properties()
     topicConfig.asScala.foreachEntry { (key, value) =>
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ConfigHandler.scala]: Is this code needed at all now?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java [hunk 2]
```diff
@@ -436,7 +375,7 @@ private Optional<Compression> getCompression() {
     }
 
     public RecordVersion recordVersion() {
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java]: Shall we remove this method? It doesn't make sense now that we don't have a log config for record version.

File: server-common/src/main/java/org/apache/kafka/server/config/ServerTopicConfigSynonyms.java [hunk 3]
```diff
@@ -49,8 +49,6 @@ public final class ServerTopicConfigSynonyms {
      * both the first and the second synonyms are configured, we will use only the value of
      * the first synonym and ignore the second.
      */
-    // Topic configs with no mapping to a server config can be found in `LogConfig.CONFIGS_WITH_NO_SERVER_DEFAULTS`
```
Comment (ijuma) [inline server-common/src/main/java/org/apache/kafka/server/config/ServerTopicConfigSynonyms.java:52]: Why did we remove this comment?

File: docs/upgrade.html [hunk 4]
```diff
@@ -61,6 +61,8 @@ <h5><a id="upgrade_400_notable" href="#upgrade_400_notable">Notable changes in 4
                         </li>
                         <li>The <code>org.apache.kafka.clients.producer.internals.DefaultPartitioner</code> and <code>org.apache.kafka.clients.producer.UniformStickyPartitioner</code> class was removed.
                         </li>
+                        <li>The <code>log.message.format.version</code> and <code>message.format.version</code> were removed.
```
Comment (ijuma) [inline docs/upgrade.html]: We should add the word `configs` before `were removed`.

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala [hunk 5]
```diff
@@ -356,12 +355,8 @@ object TestUtils extends Logging {
     props
   }
 
-  @nowarn("cat=deprecation")
   def setIbpAndMessageFormatVersions(config: Properties, version: MetadataVersion): Unit = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/utils/TestUtils.scala]: We should rename this method.

File: core/src/test/scala/integration/kafka/api/ConsumerWithLegacyMessageFormatIntegrationTest.scala [hunk 6]
```diff
@@ -113,9 +106,7 @@ class ConsumerWithLegacyMessageFormatIntegrationTest extends AbstractConsumerTes
     prop.setProperty(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "false")
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/ConsumerWithLegacyMessageFormatIntegrationTest.scala]: Shall we remove this property and comment? Also, the topic name should be updated.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. It's looking pretty good. I left a few comments.

File: coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java [hunk 7]
```diff
@@ -857,14 +858,13 @@ private void maybeAllocateNewBatch(
...
                 LogConfig logConfig = partitionWriter.config(tp);
-                byte magic = logConfig.recordVersion().value;
                 int maxBatchSize = logConfig.maxMessageSize();
                 long prevLastWrittenOffset = coordinator.lastWrittenOffset();
                 ByteBuffer buffer = bufferSupplier.get(maxBatchSize);
 
                 MemoryRecordsBuilder builder = new MemoryRecordsBuilder(
                     buffer,
-                    magic,
+                    RecordVersion.V2.value,
```
Comment (ijuma) [inline coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java]: Perhaps use `RecordBatch.CURRENT_MAGIC_VALUE` here and other similar places.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 8]
```diff
@@ -2046,7 +2046,7 @@ class ReplicaManager(val config: KafkaConfig,
 
   def getLogConfig(topicPartition: TopicPartition): Option[LogConfig] = localLog(topicPartition).map(_.config)
 
-  def getMagic(topicPartition: TopicPartition): Option[Byte] = getLogConfig(topicPartition).map(_.recordVersion.value)
+  def getMagic(topicPartition: TopicPartition): Option[Byte] = getLogConfig(topicPartition).map(_ => RecordVersion.V2.value)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: Do we still need this method?

**Review by ijuma — COMMENTED** [review]
Thanks for the updates, just a couple more comments.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 8 above)
Comment (FrankYang0529) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Thanks for review and suggestion. I try to remove `ReplicaManager#getMagic` and it looks like some related logic on this path can be removed too. Can I also remove them? Thanks.

https://github.com/apache/kafka/blob/fe56fc98fa736c79c9dcbb1f64f810065161a1cc/clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java#L93-L118

https://github.com/apache/kafka/blob/fe56fc98fa736c79c9dcbb1f64f810065161a1cc/clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java#L94-L119

Comment (ijuma) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Perhaps we can leave the changes to the records for a separate PR. One thing we have to be careful about is that old record formats may exist on disk - so the functionality to handle that needs to remain.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala [hunk 9]
```diff
@@ -1982,25 +1982,8 @@ class GroupMetadataManagerTest {
...
-    // verify the tombstones are correct and only for the expired offsets
-    val records = recordsCapture.getValue.records.asScala.toList
-    assertEquals(2, records.size)
-    records.foreach { message =>
-      assertTrue(message.hasKey)
-      assertFalse(message.hasValue)
-      val offsetKey = GroupMetadataManager.readMessageKey(message.key).asInstanceOf[OffsetKey]
-      assertEquals(groupId, offsetKey.key.group)
-      assertEquals("foo", offsetKey.key.topicPartition.topic)
-    }
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala]: Why is this no longer needed?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 10]
```diff
@@ -3120,19 +3099,16 @@ class KafkaApisTest extends Logging {
   }
 
   @Test
-  def shouldRespondWithUnsupportedMessageFormatForBadPartitionAndNoErrorsForGoodPartition(): Unit = {
-    val tp1 = new TopicPartition("t", 0)
+  def shouldRespondWithNoErrorsForGoodPartition(): Unit = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: I think we can delete this test. `shouldRespondWithUnknownTopicOrPartitionForBadPartitionAndNoErrorsForGoodPartition` seems to cover this path already.

**Comment by ijuma** [comment]
I left a couple of comments above and pushed a commit with a few improvements. Outside of the two comments, I think we're good.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 10 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Pushed a commit that does this.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala (hunk 9 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala]: I added back this code and made the test pass by returning the right partition from replica manager.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala [hunk 11]
```diff
@@ -2426,10 +2426,7 @@ class GroupMetadataManagerTest {
 
     groupMetadataManager.cleanupGroupMetadata()
 
-    verify(partition).appendRecordsToLeader(any[MemoryRecords],
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala:2429]: Also added this back (and fixed the test).

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 12]
```diff
@@ -868,56 +867,55 @@ class GroupMetadataManager(brokerId: Int,
         (removedOffsets, group.is(Dead), group.generationId)
       }
 
-    val offsetsPartition = partitionFor(groupId)
```
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:871]: The indenting of this existing code was wrong, fixed it as part of this change.

**Review by ijuma — APPROVED** [review]
LGTM with the latest updates.

**Comment by ijuma** [comment]
@FrankYang0529 If the tests pass, I'll go ahead and merge. Please take a look when you have a chance and let me know if you see any issues with my updates.

**Comment by FrankYang0529** [comment]
Hi @ijuma, thanks for reviewing and updating the PR. I think the change is good. 👍

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java [hunk 13]
```diff
@@ -223,11 +222,6 @@ public boolean renameDir(String name) {
     public void updateConfig(LogConfig newConfig) {
         LogConfig oldConfig = config;
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java:223]: we don't need this local variable now - maybe it needs a minor to cleanup :)

**Pull request merged**